compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -1552,12 +1576,72 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 }
 
                 err.emit_unless(unsized_return);
+                if ret_vec.len() > 0 {
+                    if let Some(expr) = expression {
+                        let mut addtion_err = fcx.tcx.sess.diagnostic().span_note_diag(expr.span, "the function expects a value to always be returned, but loops might run zero times");
+                        for i in 0..ret_vec.len() {
+                            //let ret_expr = fcx.tcx.hir().get(ret_vec[i]);
+                            addtion_err.span_label(
+                                fcx.tcx.hir().span(ret_vec[i]),
+                                "if the loop doesn't execute, this value would never get returned"
+                                    .to_string(),
+                            );
+                        }
+                        let loop_span = if let hir::ExprKind::Loop(_, _, _, loop_span) = expr.kind {
+                            loop_span
+                        } else {
+                            todo!()
+                        };
+                        addtion_err.span_label(
+                            loop_span,
+                            "this might have zero elements to iterate on".to_string(),
+                        );
+                        addtion_err.help("return a value for the case when the loop has zero elements to iterate on, and consider changing the return type to account for that possibility");
+                        addtion_err.emit();
```

Comment:
We could move all this code to a method that we unconditionally call, so that the happy path doesn't get harder to read, reduce indentation and so we're able to use let/else:

```suggestion
    fn note_unreachable_loop_return(
        &self,
        err: &mut DiagnoticBuilder<'_>,
        expr: hir::Expr<'tcx>,
        ret_exprs: Vec<&'tcx hir::Expr<'tcx>>
    ) {
        if ret_exprs.is_empty() {
            return;
        }
        let Some(expr) = expr else { return; }
        let hir::ExprKind::Loop(_, _, _, loop_span) = expr.kind else { return; }

        let mut span: MultiSpan = vec![loop_span].into();
        span.push_span_label(
            loop_span,
            "this might have zero elements to iterate on".to_string(),
        );
        for ret in ret_exprs {
            span.push_span_label(
                ret.span,
                "if the loop doesn't execute, this value would never get returned"
                    .to_string(),
            );
        }
        
        err.span_note(
            span,
            "the function expects a value to always be returned, but loops might run zero times",
        );

        err.help(
            "return a value for the case when the loop has zero elements to iterate on, and \
             consider changing the return type to account for that possibility"
        );
    }
```

Reply:
I've put them into the function as you suggestion, but made some minor change. Because in most cases the size of vector of return expressions should be 0. So I put calling note_unreachable_loop_return() under the if condition when visitor.ret_vec is not empty, then we can save a function call, though inliner might optimize this function call